### ISSUE TITLE
Error when parsing java/util/Map.java

### ISSUE THREAD

**Issue by kushal256** [javaparser/javaparser#174]
This is looking at jdk1.8.0_51\src.zip, at the file java.util.Map.java:

```
    public static <K extends Comparable<? super K>, V> Comparator<Map.Entry<K,V>> comparingByKey() {
        return (Comparator<Map.Entry<K, V>> & Serializable)  // line 471, col 46 is before the '>>'
            (c1, c2) -> c1.getKey().compareTo(c2.getKey());
    }
```

Here's the exception:
com.github.javaparser.ParseException: Encountered " ">" ">> "" at line 471, column 46.
Was expecting one of:
    ")" ...    "[" ...    "," ...    "[" ...    "," ...    "[" ...    "," ...    "[" ...    "," ...    "[" ...    "," ...    "[" ...    "," ...    "[" ...    "," ...
at com.github.javaparser.ASTParser.generateParseException(ASTParser.java:9049)
    at com.github.javaparser.ASTParser.jj_consume_token(ASTParser.java:8922)
    at com.github.javaparser.ASTParser.PrimaryPrefix(ASTParser.java:3037)
    at com.github.javaparser.ASTParser.PrimaryExpression(ASTParser.java:2801)
    at com.github.javaparser.ASTParser.PostfixExpression(ASTParser.java:2721)
    at com.github.javaparser.ASTParser.UnaryExpressionNotPlusMinus(ASTParser.java:2705)
    at com.github.javaparser.ASTParser.UnaryExpression(ASTParser.java:2617)
    at com.github.javaparser.ASTParser.MultiplicativeExpression(ASTParser.java:2514)
    at com.github.javaparser.ASTParser.AdditiveExpression(ASTParser.java:2477)
    at com.github.javaparser.ASTParser.ShiftExpression(ASTParser.java:2440)
    at com.github.javaparser.ASTParser.RelationalExpression(ASTParser.java:2393)
    at com.github.javaparser.ASTParser.InstanceOfExpression(ASTParser.java:2374)
    at com.github.javaparser.ASTParser.EqualityExpression(ASTParser.java:2338)
    at com.github.javaparser.ASTParser.AndExpression(ASTParser.java:2315)
    at com.github.javaparser.ASTParser.ExclusiveOrExpression(ASTParser.java:2293)
    at com.github.javaparser.ASTParser.InclusiveOrExpression(ASTParser.java:2271)
    at com.github.javaparser.ASTParser.ConditionalAndExpression(ASTParser.java:2249)
    at com.github.javaparser.ASTParser.ConditionalOrExpression(ASTParser.java:2227)
    at com.github.javaparser.ASTParser.ConditionalExpression(ASTParser.java:2207)
    at com.github.javaparser.ASTParser.Expression(ASTParser.java:2068)
    at com.github.javaparser.ASTParser.ReturnStatement(ASTParser.java:4450)
    at com.github.javaparser.ASTParser.Statement(ASTParser.java:3617)
    at com.github.javaparser.ASTParser.BlockStatement(ASTParser.java:3743)
    at com.github.javaparser.ASTParser.Statements(ASTParser.java:1555)
    at com.github.javaparser.ASTParser.Block(ASTParser.java:3680)
    at com.github.javaparser.ASTParser.MethodDeclaration(ASTParser.java:1219)
    at com.github.javaparser.ASTParser.ClassOrInterfaceBodyDeclaration(ASTParser.java:948)
    at com.github.javaparser.ASTParser.ClassOrInterfaceBody(ASTParser.java:854)
    at com.github.javaparser.ASTParser.ClassOrInterfaceDeclaration(ASTParser.java:444)
    at com.github.javaparser.ASTParser.ClassOrInterfaceBodyDeclaration(ASTParser.java:921)
    at com.github.javaparser.ASTParser.ClassOrInterfaceBody(ASTParser.java:854)
    at com.github.javaparser.ASTParser.ClassOrInterfaceDeclaration(ASTParser.java:444)
    at com.github.javaparser.ASTParser.TypeDeclaration(ASTParser.java:372)
    at com.github.javaparser.ASTParser.CompilationUnit(ASTParser.java:185)
    at com.github.javaparser.JavaParser.parse(JavaParser.java:112)
    at com.github.javaparser.JavaParser.parse(JavaParser.java:134)

**Comment by ptitjes** [comment]
Hi!
I can't find some sources for build 51. I only found sources for build 60 and I had a similar problem while parsing java/util/Comparator.java (see stack trace below). This one for sure is because of #149 (with a fix here: https://github.com/javaparser/javaparser/pull/150) as JLaTo parses it fine.

So looking at your problem, its location and the stack-trace I can only guess that it is the same problem (that is: intersection types not recognized in casts).

Would you mind try applying #150 on your local javaparser build or use branch https://github.com/ptitjes/javaparser/tree/3.0-type-hierarchy-and-annotations (an equivalent fix with a little parse time degradation) and try to parse your jdk sources again please ?

If it parses fine with either of those fixes then this issue is a duplicate of #149.

---

Parsing: /home/didier/Downloads/Tech/Software/Java/jdk1.8.0_60/src.zip
java/util/Comparator.java

com.github.javaparser.ParseException: Encountered " "&" "& "" at line 215, column 31.
Was expecting one of:
    "boolean" ... "byte" ... "char" ... "double" ... "false" ... "float" ... "int" ... "long" ... "new" ... "null" ... "short" ... "super" ... "this" ... "true" ... "void" ... <LONG_LITERAL> ... <INTEGER_LITERAL> ... <FLOATING_POINT_LITERAL> ... <CHARACTER_LITERAL> ... <STRING_LITERAL> ... <IDENTIFIER> ... "(" ... "!" ... "~" ... "++" ... "--" ... "+" ... "-" ... <IDENTIFIER> ...

```
at com.github.javaparser.ASTParser.generateParseException(ASTParser.java:9049)
at com.github.javaparser.ASTParser.jj_consume_token(ASTParser.java:8922)
at com.github.javaparser.ASTParser.UnaryExpression(ASTParser.java:2621)
at com.github.javaparser.ASTParser.MultiplicativeExpression(ASTParser.java:2514)
at com.github.javaparser.ASTParser.AdditiveExpression(ASTParser.java:2477)
at com.github.javaparser.ASTParser.ShiftExpression(ASTParser.java:2440)
at com.github.javaparser.ASTParser.RelationalExpression(ASTParser.java:2429)
at com.github.javaparser.ASTParser.InstanceOfExpression(ASTParser.java:2374)
at com.github.javaparser.ASTParser.EqualityExpression(ASTParser.java:2338)
at com.github.javaparser.ASTParser.AndExpression(ASTParser.java:2315)
at com.github.javaparser.ASTParser.ExclusiveOrExpression(ASTParser.java:2293)
at com.github.javaparser.ASTParser.InclusiveOrExpression(ASTParser.java:2271)
at com.github.javaparser.ASTParser.ConditionalAndExpression(ASTParser.java:2249)
at com.github.javaparser.ASTParser.ConditionalOrExpression(ASTParser.java:2227)
at com.github.javaparser.ASTParser.ConditionalExpression(ASTParser.java:2207)
at com.github.javaparser.ASTParser.Expression(ASTParser.java:2068)
at com.github.javaparser.ASTParser.PrimaryPrefix(ASTParser.java:3015)
at com.github.javaparser.ASTParser.PrimaryExpression(ASTParser.java:2801)
at com.github.javaparser.ASTParser.PostfixExpression(ASTParser.java:2721)
at com.github.javaparser.ASTParser.UnaryExpressionNotPlusMinus(ASTParser.java:2705)
at com.github.javaparser.ASTParser.UnaryExpression(ASTParser.java:2617)
at com.github.javaparser.ASTParser.MultiplicativeExpression(ASTParser.java:2514)
at com.github.javaparser.ASTParser.AdditiveExpression(ASTParser.java:2477)
at com.github.javaparser.ASTParser.ShiftExpression(ASTParser.java:2440)
at com.github.javaparser.ASTParser.RelationalExpression(ASTParser.java:2393)
at com.github.javaparser.ASTParser.InstanceOfExpression(ASTParser.java:2374)
at com.github.javaparser.ASTParser.EqualityExpression(ASTParser.java:2338)
at com.github.javaparser.ASTParser.AndExpression(ASTParser.java:2315)
at com.github.javaparser.ASTParser.ExclusiveOrExpression(ASTParser.java:2293)
at com.github.javaparser.ASTParser.InclusiveOrExpression(ASTParser.java:2271)
at com.github.javaparser.ASTParser.ConditionalAndExpression(ASTParser.java:2249)
at com.github.javaparser.ASTParser.ConditionalOrExpression(ASTParser.java:2227)
at com.github.javaparser.ASTParser.ConditionalExpression(ASTParser.java:2207)
at com.github.javaparser.ASTParser.Expression(ASTParser.java:2068)
at com.github.javaparser.ASTParser.ReturnStatement(ASTParser.java:4450)
at com.github.javaparser.ASTParser.Statement(ASTParser.java:3617)
at com.github.javaparser.ASTParser.BlockStatement(ASTParser.java:3743)
at com.github.javaparser.ASTParser.Statements(ASTParser.java:1555)
at com.github.javaparser.ASTParser.Block(ASTParser.java:3680)
at com.github.javaparser.ASTParser.MethodDeclaration(ASTParser.java:1219)
at com.github.javaparser.ASTParser.ClassOrInterfaceBodyDeclaration(ASTParser.java:948)
at com.github.javaparser.ASTParser.ClassOrInterfaceBody(ASTParser.java:854)
at com.github.javaparser.ASTParser.ClassOrInterfaceDeclaration(ASTParser.java:444)
at com.github.javaparser.ASTParser.TypeDeclaration(ASTParser.java:372)
at com.github.javaparser.ASTParser.CompilationUnit(ASTParser.java:185)
at com.github.javaparser.JavaParser.parse(JavaParser.java:112)
at org.jlato.test.BenchmarkIT$1.parse(BenchmarkIT.java:30)
at org.jlato.test.BenchmarkIT.doParse(BenchmarkIT.java:117)
at org.jlato.test.BenchmarkIT.parseFrom(BenchmarkIT.java:80)
at org.jlato.test.BenchmarkIT.parseJDK(BenchmarkIT.java:58)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
```

**Comment by kushal256** [comment]
Yes, you are correct.

I used https://github.com/ptitjes/javaparser/tree/3.0-type-hierarchy-and-annotations and that fixed the problem ;)

Thx very much!  I'll close this issue.

**Comment by ptitjes** [comment]
You're welcome!

Duplicate of #149.

**Comment by ptitjes** [comment]
@krushal256 Please note that I won't maintain the fixes in https://github.com/ptitjes/javaparser/tree/3.0-type-hierarchy-and-annotations nor will they make their way to javaparser master due to the license change of javaparser.

So I advise you to either use the fix in #150 made by a contributor that agrees with the new license scheme, or to use https://github.com/ptitjes/jlato which correctly parses the jdk.
